### PR TITLE
curvefs fix load copyset stuck heartbeat #1468

### PR DESCRIPTION
**Pull request by cw123**
### What problem does this PR solve?

Issue Number: #1468

Problem Summary: curvefs fix load copyset stuck heartbeat #1468

### What is changed and how it works?

When copyset is loading, it will stuck heartbeat when get partition info list. I make some changes.
1. When copyset is loading, heartbeat do not send partition info in copyset.
2. When copyset get partition, use try lock instead of lock.
3. Do not use heartbeat partition num to update topo.

### CODE REVIEW

File: curvefs/src/metaserver/heartbeat.cpp [hunk 1]
```diff
@@ -185,9 +185,20 @@ void Heartbeat::BuildCopysetInfo(curvefs::mds::heartbeat::CopySetInfo *info,
...
-        info->add_partitioninfolist()->CopyFrom(it);
+    if (!isLoading) {
+        std::list<PartitionInfo> partitionInfoList;
+        bool ret = copyset->GetPartitionInfoList(&partitionInfoList);
+        if (ret) {
+            for (auto& it : partitionInfoList) {
+                info->add_partitioninfolist()->CopyFrom(it);
+            }
+        } else {
+            info->set_iscopysetloading(true);
```
Comment (wu-hanqing) [inline curvefs/src/metaserver/heartbeat.cpp:200]: log a warning message?

File: curvefs/src/metaserver/metastore.cpp [hunk 2]
```diff
@@ -175,16 +175,28 @@ MetaStatusCode MetaStoreImpl::DeletePartition(
...
+        rwLock_.Unlock();
+        return true;
+    } else {
+        LOG(WARNING) << "metastore GetPartitionInfoList fail, it fail to get"
+                        " the rwLock_";
+        return false;
     }
-    return partitionInfoList;
+
+    return true;
```
Comment (wu-hanqing) [inline curvefs/src/metaserver/metastore.cpp:199]: it's redundant

File: curvefs/proto/heartbeat.proto [hunk 3]
```diff
@@ -37,7 +37,8 @@ message CopySetInfo {
     required common.Peer leaderPeer = 5;
     repeated common.PartitionInfo partitionInfoList = 6;
     optional ConfigChangeInfo configChangeInfo = 7;
-    // required CopySetState state = 8;
+    optional bool isCopysetLoading = 8;
```
Comment (wu-hanqing) [inline curvefs/proto/heartbeat.proto:40]: this optional field will introduce problems when the version of MDS and Metaserver are mismatching

File: curvefs/src/metaserver/copyset/copyset_node.cpp [hunk 4]
```diff
@@ -351,10 +352,12 @@ int CopysetNode::on_snapshot_load(braft::SnapshotReader* reader) {
     LOG(INFO) << "Copyset " << name_ << " begin to load snapshot from '"
               << reader->get_path() << "'";
 
+    isLoading_.store(true, std::memory_order_release);
```
Comment (wu-hanqing) [inline curvefs/src/metaserver/copyset/copyset_node.cpp:355]:
```cpp
struct CopysetLoadingGuard {
    CopysetLoadingGuard(std::atomic& flag) : flag(flag) {
           flag.store(true, ...);
    }

    ~CopysetLoadingGuard() {
          flags.stroe(false, ....);
    }


    std::atomic& flags;
};
```

in this way, you don't have to manually set `isLoading_` to false in every return point.

File: curvefs/src/mds/topology/topology_item.h [hunk 5]
```diff
@@ -548,7 +550,7 @@ class CopySetInfo {
 
     uint64_t GetPartitionNum() const { return partitionNum_; }
 
-    void SetPartitionNum(u_int64_t number) { partitionNum_ = number; }
+    void SetPartitionNum(uint64_t number) { partitionNum_ = number; }
```
Comment (SeanHai) [inline curvefs/src/mds/topology/topology_item.h:553]: The partitionNum_ field is needn't anymore, remove it and related functions.

File: curvefs/src/metaserver/copyset/copyset_node.cpp (hunk 4 above)
Comment (SeanHai) [reply curvefs/src/metaserver/copyset/copyset_node.cpp:355]: agree

File: curvefs/src/metaserver/copyset/copyset_node.cpp [hunk 6]
```diff
@@ -540,8 +544,22 @@ void CopysetNode::ListPeers(std::vector<Peer>* peers) const {
...
 
-std::list<PartitionInfo> CopysetNode::GetPartitionInfoList() {
-    return metaStore_->GetPartitionInfoList();
+// if copyset is loading, return false;
+// if copyset is not loading, and metastore returns fales, retry;
+// if copyset is not loading, and metastore returns true, return true and get
+// partition info list success.
+bool CopysetNode::GetPartitionInfoList(
+                    std::list<PartitionInfo> *partitionInfoList) {
+    while (true) {
```
Comment (SeanHai) [inline curvefs/src/metaserver/copyset/copyset_node.cpp:553]: Does the infinite loop logic is ok when metaStore_->GetPartitionInfoList failed？

Comment (cw123) [reply curvefs/src/metaserver/copyset/copyset_node.cpp:553]: Under normal circumstances, one retry can be successful.

File: curvefs/src/metaserver/heartbeat.cpp (hunk 1 above)
Comment (cw123) [reply curvefs/src/metaserver/heartbeat.cpp:200]: done

File: curvefs/src/metaserver/metastore.cpp (hunk 2 above)
Comment (cw123) [reply curvefs/src/metaserver/metastore.cpp]: done

File: curvefs/proto/heartbeat.proto (hunk 3 above)
Comment (cw123) [reply curvefs/proto/heartbeat.proto:40]: done

File: curvefs/src/metaserver/copyset/copyset_node.cpp (hunk 4 above)
Comment (cw123) [reply curvefs/src/metaserver/copyset/copyset_node.cpp]: done

File: curvefs/src/mds/topology/topology_item.h (hunk 5 above)
Comment (cw123) [reply curvefs/src/mds/topology/topology_item.h]: done

**Comment by cw123** [comment]
recheck

File: curvefs/src/mds/topology/topology_item.h [hunk 7]
```diff
@@ -497,6 +497,7 @@ class CopySetInfo {
           epoch_(v.epoch_),
           peers_(v.peers_),
           partitionNum_(v.partitionNum_),
+          partitionIds_(v.partitionIds_),
```
Comment (SeanHai) [inline curvefs/src/mds/topology/topology_item.h:500]: The partitionNum_ is not needed. In GetPartitionNum can return partitionIds_.size().

Comment (cw123) [reply curvefs/src/mds/topology/topology_item.h:500]: If this is removed, the software will have problems with backward compatibility.

Comment (cw123) [reply curvefs/src/mds/topology/topology_item.h:497]: done, i have removed partitionNum_.

**Review by SeanHai — APPROVED** [review]

**Review by wu-hanqing — APPROVED** [review]

**Pull request merged**